File: CrackingTheCodingInterview/TechnicalQuestions/LowestCommonAncestor.java
package CrackingTheCodingInterview.TechnicalQuestions;

public class LowestCommonAncestor {
    public static void main(String[] args) {
        
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return null;
        if (p == root || q == root) return root;
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if (left != null && right != null) {
            return root;
        }
        if (left == null && right == null) {
            return null;
        }
        return left != null ? left : right;
    }
}

File: CrackingTheCodingInterview/Udemy/LowestCommonAncestor.js
var lowestCommonAncestor = function(root, p, q) {
    if (!root) return null;
    if (root == p || root == q) return root;
    
    let left = lowestCommonAncestor(root.left, p, q);
    let right = lowestCommonAncestor(root.right, p, q);
    
    if (left && right) return root;
    
    return left != null ? left : right;
};

var lowestCommonAncestorBST = function(root, p, q) {
    if (!root) return null;
    if (root.val > p.val && root.val > q.val) {
        return lowestCommonAncestor(root.left, p, q)
    }
    
    if (root.val < p.val && root.val < q.val) {
        return lowestCommonAncestor(root.right, p, q);
    }
    
    return root;
};